text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,229 @@
+- Tracking: (leave this empty)
+
+# Custom Model Class RFC
+
+## Summary
+
+This RFC is a follow-up RFC for #293 RecordData and replaces [https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor](https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor)
+
+Create a way for addons and user to define their own implemnentation of a model clas. Adds an `instantiateRecord` method to `Store` which would allow addons and ED itself to offer a clean replacement the default DS.Model with a custom Model class. 
+
+## Motivation
+
+- Allowing Addons to experiment with replacing the default DS.Model implementation ([https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor](https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor))
+- Currently Ember Data conflates schema information and record implementation together as part of properties defined on `DS.Model`.
+DS.Model both defines the schema, that is the attributes and relationships present on a type, and the model implementation, that is the
+actual class which is exposed to the app containing those properties. This has several disadvantages:
+- it makes it hard to optimize the schema information, and ties to a definition of a JS class
+- it makes it hard to statically analyze the schema
+- it requires a 1-1 correspondence between the number of backing types in the system and the number of JS classes, resulting
+in an app sometimes having to ship a much higher number of classes than it otherwise would have.
+
+This RFC represents the first step in separating the record implementation from schema information.
+
+## Detailed design
+
+After the work in [Record Data Errors RFC](https://github.com/emberjs/rfcs/pull/465),  [Record Data State RFC](https://github.com/emberjs/rfcs/pull/463), and [Request State Service RFC](https://github.com/emberjs/rfcs/pull/466) we have enough coverage in public apis that we can implement the entire DS.Model class with just a few changes to Store APIs. 
+
+We will need five separate changes: 
+
+- A way to instantiate custom Records and for them to be able to access underlying record data
+- A way for those records be notified of changes in Record Data
+- A way to access schema information not from the Record Class
+- Replacement of few store methods that take DS.Models
+- Deprecate passing DS.Model classes to adapter/serializers/exposing them on Snapshots
+
+## Instantiating and destroying custom Records
+
+First we need to expose a method on the store for instantiating a record, a hook to be notified when a record is being destroyed for potential cleanup and a way for the record to get notified when RecordData values change:
```

Comment:
This would probably be easier to grok as a list:

```suggestion
We need to make the following changes:

* expose a method on the store for instantiating a record
* add a hook to be notified when a record is being destroyed for potential cleanup
* a way for the record to get notified when RecordData values change

The following interface shows the interface of these methods:
```